web/htdocs/assets/js/codesearch_ui.js

```diff
@@ -72,76 +147,78 @@ var Match = Backbone.Model.extend({
-    // Context_before is actually in reverse order...
-    for (i = 0; i < this.model.get('context_before').length; i ++) {
-      var cells = [];
-      cells.push(h.td({cls: 'lno', line_number: matchlno-(i+1)},[]));
-      cells.push(h.td({cls: 'line'},[this.model.get('context_before')[i]]));
-      lines.unshift(h.tr({cls: 'row'},cells));
-    }
 
-    var matchCells = [];
-    // Bringing lno up to the actual match's line number
-    var line = this.model.get('line');
-    var bounds = this.model.get('bounds');
-    var pieces = [line.substring(0, bounds[0]),
-                  line.substring(bounds[0], bounds[1]),
-                  line.substring(bounds[1])];
-    matchCells.push(h.td({cls: 'lno matchline', line_number: matchlno}, []));
-    matchCells.push(h.td({cls: 'line matchline'}, [
-                  pieces[0],
-                  h.span({cls: 'matchstr'}, [pieces[1]]),
-                  pieces[2]
-                ]));
-    lines.push(h.tr({cls: 'matchrow row'},matchCells));
-
-    for (i = 0; i < this.model.get('context_after').length; i ++) {
-      var cells = [];
-      cells.push(h.td({cls: 'lno', line_number: matchlno+(i+1)},[]));
-      cells.push(h.td({cls: 'line'},[this.model.get('context_after')[i]]));
-      lines.push(h.tr({cls: 'row'},cells));
+  /** Prepare the matches for rendering by clipping the context of matches to avoid duplicate
+   *  lines being displayed in the search results.
+   *
+   * This function operates under these assumptions:
+   * - The matches are all for the same file
+   * - Two matches cannot have the same line number
+   */
+  process_context_overlaps: function() {
+    if(!(this.matches) || this.matches.length < 2) {
+      return; // We don't have overlaps unless we have at least two things
     }
 
-    var tree = this.model.get('tree');
-    var version = this.model.get('version');
-    var repoLabel = [
-      tree ? (tree + ":") : "",
-      shorten(version),
-      ":",
-      this.model.get('path'),
-      " - line " + matchlno + "" ];
-    var url = this.model.url();
-    if (url !== null) {
-      repoLabel = [ h.a({href: this.model.url()}, repoLabel) ];
+    // NOTE: The logic below requires matches to be sorted by line number.
+    this.matches.sort(function(a, b) {
+      return a.get('lno') - b.get('lno');
+    });
+
+    for(var i = 1, len = this.matches.length; i < len; i++) {
+      var previous_match = this.matches[i - 1], this_match = this.matches[i];
+      var last_line_of_prev_context = previous_match.get('lno') + previous_match.get('context_after').length;
+      var first_line_of_this_context = this_match.get('lno') - this_match.get('context_before').length;
+      var num_intersecting_lines = (last_line_of_prev_context - first_line_of_this_context) + 1;
+      if(num_intersecting_lines >= 0) {
+        // The matches are intersecting or share a boundary.
+        // Split the context between the previous match and this one. Uneven splits should leave
```

Comment:
I think this is the culprit for the bug that you are seeing. This splitting algorithm doesn't work all the time if the two matches do not have a balanced number of context lines. This typically occurs at the beginning or end of the file. For example:

previous_match = {'lno': 1, 'context_before':[], 'context_after':[2, 3, 4]}
this_match = {'lno': 2, 'context_before':[1], 'context_after': [3, 4, 5]}

The algorithm will try to split it even and remove 2 lines from previous_match and 2 lines from this_match. Resulting in:
previous_match = {'lno': 1, 'context_before':[], 'context_after':[2]}
this_match = {'lno': 2, 'context_before':[], 'context_after': [3, 4, 5]}